cenpy/legacy_tiger.py

```diff
@@ -0,0 +1,234 @@
+            except:
+                module.poly_line_links[rpoly_id] = [tlid]
+
+            link_count = link_count + 1
+
+            feat.Destroy()
+
+            feat = link_layer.GetNextFeature()
+
+        # Process all polygon features.
+
+        feat = poly_layer.GetNextFeature()
+        tile_ref_field = feat.GetFieldIndex('MODULE')
+        polyid_field = feat.GetFieldIndex('POLYID')
+
+        degenerate_count = 0
+
+        while feat is not None:
+            module = modules_hash[feat.GetField(tile_ref_field)]
+            polyid = feat.GetField(polyid_field)
+
+            tlid_list = module.poly_line_links[polyid]
+
+            link_coll = ogr.Geometry(type=ogr.wkbGeometryCollection)
+            for tlid in tlid_list:
+                geom = module.lines[tlid]
+                link_coll.AddGeometry(geom)
+
+            try:
+                poly = ogr.BuildPolygonFromEdges(link_coll)
+
+                if poly.GetGeometryRef(0).GetPointCount() < 4:
+                    degenerate_count = degenerate_count + 1
+                    poly.Destroy()
+                    feat.Destroy()
+                    feat = poly_layer.GetNextFeature()
+                    continue
+
+                feat2 = ogr.Feature(feature_def=shp_layer.GetLayerDefn())
+
+                for fld_index in range(poly_field_count):
+                    feat2.SetField(fld_index, feat.GetField(fld_index))
+
+                feat2.SetGeometryDirectly(poly)
+
+                shp_layer.CreateFeature(feat2)
+                feat2.Destroy()
+
+            except:
+                warn('BuildPolygonFromEdges failed.')
+
+            feat.Destroy()
+
+            feat = poly_layer.GetNextFeature()
+
+        if degenerate_count:
+            warn('Discarded %d degenerate polygons.' % degenerate_count)
+
+        # Cleanup
+
+        shp_ds.Destroy()
+        shp_ds = None
+        ds.Destroy()
+        ds = None
+
+        # build a fully-qualified fips code and dissolve on it to create tract geographies
+        gdf = gpd.read_file(outfile)
+
+        if "CTBNA90" in gdf.columns:
+
+            gdf = gdf.rename(columns={"CTBNA90": 'TRACT', "BLK90": "BLOCK"})
+
+        gdf['STATE'] = gdf['STATE'].astype(str).str.rjust(2, "0")
+        gdf['COUNTY'] = gdf['COUNTY'].astype(str).str.rjust(3, "0")
+        gdf['TRACT'] = gdf['TRACT'].astype(str).str.rjust(6, "0")
+        gdf['BLOCK'] = gdf['BLOCK'].astype(str).str.rjust(4, "0")
+        gdf['fips'] = gdf.STATE + gdf.COUNTY + gdf.TRACT
+        if self.geom == 'block':
+            gdf['fips'] += gdf.BLOCK
+
+        gdf = gdf.dropna(subset=['fips'])
+        gdf.geometry = gdf.buffer(0)
+        gdf = gdf.dissolve(by='fips')
+        gdf.reset_index(inplace=True)
+
+        shp_driver.DeleteDataSource(outfile)
+
+        return gdf
+
+    def query(self):
+
+        with urlopen(self.url) as response:
+            with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
```

Comment:
the segfaults I was encountering earlier were due to improper handling of these temp files... They were getting garbage collected before GDAL finished processing them or something. I was able to get this function working by just (over)writing out each of the necessary files to disk and deleting the final shp at the end of the code

the working script is [here](https://gist.github.com/knaaptime/d1be08c34d433d08d7d2cf43b7ef96a1). I'd be happy to update this PR to include these changes, but i'm not sure about best practices for writing an intermediate file to disk as part of a package

Reply:
> not sure about best practices for writing an intermediate file to disk as part of a package

that makes two of us. 

I'm fine with using `tempfile` given it's in the standard library, and you're probably using it right here. I'll merge the linked version.